File: src/main/java/io/github/m/Pair.java
package io.github.m;

/**
 * M wrapper class for pairs.
 */
public final class Pair implements Value {
    /**
     * The first value of the pair.
     */
    public final Value first;

    /**
     * The second value of the pair.
     */
    public final Value second;

    public Pair(Value first, Value second) {
        this.first = first;
        this.second = second;
    }

    @Override
    public String toString() {
        return "(" + first + ", " + second + ")";
    }

    /**
     * Applies the argument to the first and second elements of this pair.
     */
    @Override
    public Value invoke(Value arg) {
        return arg.invoke(first, second);
    }

    /**
     * Converts a value to a pair.
     */
    public static Pair from(Value value) {
        if (value instanceof Pair) {
            return (Pair) value;
        } else {
            return new Pair(value.invoke(Bool.TRUE), value.invoke(Bool.FALSE));
        }
    }

    @MField(name = "pair")
    public static Value pair = new Value.Impl2("pair", Pair::new);

    @MField(name = "first")
    public static Value $first = new Value.Impl1("first", pair -> Pair.from(pair).first);

    @MField(name = "second")
    public static Value $second = new Value.Impl1("second", pair -> Pair.from(pair).second);
}

File: src/main/kotlin/io/github/m/Pair.kt
package io.github.m

/**
 * Class representing an M pair.
 */
data class Pair(val first: Value, val second: Value) : Value {
    override fun invoke(arg: Value): Value = when (arg) {
        Bool.TRUE -> first
        Bool.FALSE -> second
        else -> arg(first, second)
    }

    companion object {
        fun from(value: Value) = value as? Pair ?: Pair(value(Bool.TRUE), value(Bool.FALSE))
    }

    /**
     * M pair definitions.
     */
    @Suppress("unused")
    object Definitions {
        @MField(name = "pair")
        @JvmField
        val pair: Value = Value.Impl2 { first, second -> Pair(first, second) }

        @MField(name = "first")
        @JvmField
        val first: Value = Value.Impl1 { pair -> from(pair).first }

        @MField(name = "second")
        @JvmField
        val second: Value = Value.Impl1 { pair -> from(pair).second }
    }
}